### PR TITLE
Update to ruff, update workflow versions

### LINKED ISSUE #1854: [Feature]: Migrate from setup.py to pyproject.toml
### What would you like to see added to PyNWB?

Migrate from setup.py to pyproject.toml

### Is your feature request related to a problem?

_No response_

### What solution would you like?

Migrate from setup.py to pyproject.toml

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you checked the [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this change was not already [requested](https://github.com/NeurodataWithoutBorders/pynwb/issues)?

### PR DESCRIPTION
**Pull request by mavaylon1**
## Motivation

What was the reasoning behind this change? Please explain the changes briefly.
Fix #1854 
- [x] Finish updating tox.ini to support 3.12
- [x] Update coverage and codespell 
- [x] Migrate to use hatch and hatch-vcs
- [x] Fix ros3 failure
- [x] Update gallery tests to drop imp module for importlib.utils due to python 3.12
- [x] Add ruff and drop flake8
## How to test the behavior?
```
Show how to reproduce the new behavior (can be a bug fix or a new feature)
```

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by mavaylon1** [comment]
Note:
- add ignore points for ruff
- check coverage issue
-

**Comment by mavaylon1** [comment]
`AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?`
It seems pkgutil deprecated this method on python 3.12 (https://stackoverflow.com/questions/77364550/attributeerror-module-pkgutil-has-no-attribute-impimporter-did-you-mean)

This does not happen in hdmf or hdmf-zarr (why here?)

**Comment by mavaylon1** [comment]
@rly could you take a look at the codecov patch? I think we can ignore the issue in patch since the project coverage is passing. The patch itself is pointing to import changes I made for getting the version and also style changes in icephys.py that were already untested.

**Comment by mavaylon1** [comment]
I'm going to give this a pass through today prior to review requesting. @rly

**Comment by rly** [comment]
> I think we can ignore the issue in patch since the project coverage is passing. The patch itself is pointing to import changes I made for getting the version and also style changes in icephys.py that were already untested.

I agree. That's all fine.

File: tox.ini [hunk 1]
```diff
@@ -24,41 +24,13 @@ commands =
...
 
-# Env to create coverage report locally
-[testenv:localcoverage]
-basepython = python3.11
-commands =
-    python -m coverage run test.py --pynwb
-    coverage html -d tests/coverage/htmlcov
-
-# Test with python 3.11; pinned dev and optional reqs
-[testenv:py311-optional]
```
Comment (mavaylon1) [inline tox.ini:35]: Removed since we do not have an optional requirements file.

File: tox.ini [hunk 2]
```diff
@@ -24,41 +24,13 @@ commands =
     python -m pip list
     python test.py -v
 
-# Env to create coverage report locally
-[testenv:localcoverage]
```
Comment (mavaylon1) [inline tox.ini:28]: Removed since this does not exist in hdmf and we can look at the report on codecov itself.

File: tox.ini [hunk 3]
```diff
@@ -24,41 +24,13 @@ commands =
...
+basepython = python3.12
 install_command =
     python -m pip install -U {opts} {packages}
 deps =
     -rrequirements-dev.txt
-    ; -rrequirements-opt.txt
-commands = {[testenv]commands}
-
-# Test with python 3.11; pinned dev and optional reqs; upgraded, pre-release run reqs
-[testenv:py311-prerelease]
```
Comment (mavaylon1) [inline tox.ini:55]: Removed since this is the same as other test environments prior.

File: tox.ini [hunk 4]
```diff
@@ -91,29 +63,22 @@ commands = {[testenv:build]commands}
 basepython = python3.11
 commands = {[testenv:build]commands}
 
-[testenv:build-py311-optional]
```
Comment (mavaylon1) [inline tox.ini:94]: Removed since we do not have optional requirements file.

File: docs/source/install_developers.rst [hunk 5]
```diff
@@ -93,31 +93,19 @@ Run tests
...
+    pip install -r requirements.txt -r requirements-dev.txt
+    pip install -e .
+    tox
 
 For debugging it can be useful to keep the intermediate NWB files created by
 the tests. To keep these files create the environment variables
 ``CLEAN_NWB``/``CLEAN_HDMF`` and set them to ``1``.
 
 
-Following PyNWB Style Guide
```
Comment (mavaylon1) [inline docs/source/install_developers.rst:109]: Removed since this section is not in the hdmf install_developers.rst

Comment (rly) [reply docs/source/install_developers.rst:109]: That's fine. I think I moved it to https://hdmf.readthedocs.io/en/stable/contributing.html in HDMF. Could you update the corresponding page in PyNWB as well (if it needs an update)?

Comment (mavaylon1) [reply docs/source/install_developers.rst:109]: @rly already done!

**Review by rly — DISMISSED** [review]

**Review by rly — APPROVED** [review]

**Comment by rly** [comment]
Looks great. Thank you for taking care of this much needed infrastructure update @mavaylon1

**Pull request merged**